ui/components/src/Table/TableCell.tsx

```diff
@@ -0,0 +1,127 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { TableCell as MuiTableCell, styled, TableCellProps as MuiTableCellProps, Box, useTheme } from '@mui/material';
+import { useEffect, useRef } from 'react';
+import { TableDensity, getTableCellLayout } from './model/table-model';
+
+const StyledMuiTableCell = styled(MuiTableCell)(({ theme }) => ({
+  padding: 0,
+  backgroundColor: 'inherit',
+
+  '&.MuiTableCell-head': {
+    // Important to avoid scrolling behind the header showing through.
+    backgroundColor: theme.palette.background.paper,
+  },
+  '&:focus-visible': {
+    outline: `solid 1px ${theme.palette.primary.main}`,
+    // Move inward a little to avoid getting cut off when focusing on items
+    // at the edge of the table.
+    outlineOffset: '-1px',
+    borderRadius: 0,
+  },
+}));
+
+export interface TableCellProps extends Omit<MuiTableCellProps, 'tabIndex'> {
+  density: TableDensity;
+
+  /**
+   * How the cell should behave related to focus.
+   * - `trigger-focus`: the cell should be auto-focused when it renders.
+   * - `focus-next`: the cell should have tabindex="0", so that it will be
+   *   focused the next time someone tabs with a keyboard.
+   * - `none`: the cell should have tabindex="-1", so it is not focused by
+   *   keyboard interactions at this time.
+   */
+  focusState?: 'trigger-focus' | 'focus-next' | 'none';
+  onFocusTrigger?: (e: React.MouseEvent<HTMLTableCellElement> | React.KeyboardEvent<HTMLTableCellElement>) => void;
+}
+
+export function TableCell({
+  children,
+  density,
+  variant,
+  width,
+  focusState = 'none',
+  onFocusTrigger,
+  ...otherProps
+}: TableCellProps) {
+  const theme = useTheme();
+
+  const elRef = useRef<HTMLTableCellElement>();
+
+  const isHeader = variant === 'head';
+
+  useEffect(() => {
+    if (focusState === 'trigger-focus' && elRef.current) {
+      elRef.current.focus();
+    }
+  }, [focusState]);
+
+  const handleFocus: React.FocusEventHandler<HTMLTableCellElement> = (e) => {
+    // From https://zellwk.com/blog/keyboard-focusable-elements/
+    const nestedFocusTarget = e.currentTarget?.querySelector<HTMLElement>(
+      'a[href], button, input, textarea, select, details'
+    );
+    if (nestedFocusTarget) {
+      // If the cell has a focusable child, focus it instead. Mostly used for
+      // checkbox cells, but could have other uses.
+      nestedFocusTarget.focus();
+    }
+  };
+
+  const handleInteractionFocusTrigger: TableCellProps['onFocusTrigger'] = (e) => {
+    // We use `onClick` and `onKeyUp` events instead of `onFocus` because of
+    // some ordering issues with when the browser calls events and how this
+    // plays with the triggering of focus with keyboard interactions.
+    // These report that a focus event happened, so we can adjust the current
+    // tabindex and focuses to the right cell.
+    onFocusTrigger?.(e);
+  };
```

Comment:
These are some of the gnarlier bits of keyboard interaction code I'm not in love with. However, it's working, and I *really* want to move on with getting other bits of the table legend working, so I'd love to leave it as is and consider improvements in the future unless someone finds serious bugs in the interactions.